sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala

```diff
@@ -407,10 +407,13 @@ object ScalaReflection extends ScalaReflection {
       val externalDataType = dataTypeFor(elementType)
       val Schema(catalystType, nullable) = silentSchemaFor(elementType)
       if (isNativeType(catalystType)) {
-        NewInstance(
-          classOf[GenericArrayData],
-          input :: Nil,
-          dataType = ArrayType(catalystType, nullable))
+        expressions.If(
+          IsNull(input),
+          expressions.Literal.create(null, ArrayType(catalystType, nullable)),
+          NewInstance(
+            classOf[GenericArrayData],
+            input :: Nil,
+            dataType = ArrayType(catalystType, nullable)))
```

Comment:
A simple idea is to set `propafateNull = true` for `NewInstance`, and it's also fixed in https://github.com/apache/spark/pull/10443 which makes `false` as default value for `propagateNull`.